### PR TITLE
ColorJitter Enhancement

### PR DESCRIPTION
**Pull request by yaox12**
Set custom min and max bounds by inputing a tuple for each parameter in ColorJitter.

See issue #545

### CODE REVIEW

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -754,20 +758,53 @@ def get_params(brightness, contrast, saturation, hue):
             saturation in a random order.
         """
         transforms = []
-        if brightness > 0:
+        brightness_factor = None
+        if isinstance(brightness, float) and brightness >= 0:
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: can you make this conditional into a helper method in this class? It has a lot of redundancy and if we ever want to update something, it gets more complicated.
Something in the lines of

```python
def _sample_from(self, value, name):
    factor = None
    if isinstance(value, numbers.Number) and value >= 0:
        ...
    elif isinstance(value, (tuple, list) len(value) == 2:
        ..
    else:
        raise TypeError('{} should be ...').format(name)
```

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -754,20 +758,53 @@ def get_params(brightness, contrast, saturation, hue):
...
             brightness_factor = random.uniform(max(0, 1 - brightness), 1 + brightness)
+        elif isinstance(brightness, tuple) and len(brightness) == 2:
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: I think it might be safer to consider tuples and lists

**Comment by yaox12** [comment]
I've improve it according to your advices. Please review it. @fmassa

File: torchvision/transforms/transforms.py [hunk 3]
```diff
@@ -753,21 +757,39 @@ def get_params(brightness, contrast, saturation, hue):
             Transform which randomly adjusts brightness, contrast and
             saturation in a random order.
         """
+        def _sample_from(value, name):
+            factor = None
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: when do we have `factor==None` with the new code? Is this still needed?

**Comment by vfdev-5** [comment]
@yaox12 just wonder in which case the following conditions are met like 
```python
if brightness_factor is not None:
```
in `_sample_from` I think there is no possibility to return `None`...

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -753,21 +757,39 @@ def get_params(brightness, contrast, saturation, hue):
...
+            if isinstance(value, numbers.Number) and value >= 0:
+                if name == 'hue':
+                    factor = random.uniform(-value, value)
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: I'd prefer not to special-case for the `hue` value. What about something like the following:

```python
def _sample_from(value, name, center=1, clip_first_on_zero=True):
    if isinstance(value, numbers.Number) and value >= 0:
        value = [center - value, center + value]
    if clip_first_on_zero:
        value[0] = max(value[0], 0)
    factor = random.uniform(value[0], value[1])
```

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -753,21 +757,39 @@ def get_params(brightness, contrast, saturation, hue):
...
+                raise TypeError('{} should be non negative float or tuple of float (min, max).'.format(name))
+            return factor
+
         transforms = []
-        if brightness > 0:
-            brightness_factor = random.uniform(max(0, 1 - brightness), 1 + brightness)
+
+        brightness_factor = _sample_from(brightness, 'brightness')
+        if brightness_factor is not None:
             transforms.append(Lambda(lambda img: F.adjust_brightness(img, brightness_factor)))
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: I have the impression that now we always append the transform, even if brightness is 0, which is a performance regression compared to the previous implementation.

**Comment by yaox12** [comment]
In the previous code, if `brightness = 0`, then nothing would happen. In my code, it will call `F.adjust_brightness(img, 1)`, though it will return `img` without any change.  
I would like to make this case return `None`, but I find it will bring more `if`. Do you have any elegant solution? @vfdev-5

**Comment by vfdev-5** [comment]
@yaox12 Maybe something like in the previous code could help. Previously, if brightness value is 0 than ignore it, we can replace it with None and check as previously : 
```python
if saturation is not None:
    saturation_factor = _sample_from(saturation, 'saturation')
    transforms.append(Lambda(lambda img: F.adjust_saturation(img, saturation_factor)))
```
HTH

File: torchvision/transforms/transforms.py (hunk 5 above)
Comment (yaox12) [reply torchvision/transforms/transforms.py]: The current implementation doesn't append the transform when brightness is 0. See [Line 757](https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py#L757), so should we keep this or append the transform whatever the value is?

File: torchvision/transforms/transforms.py (hunk 4 above)
Comment (yaox12) [reply torchvision/transforms/transforms.py]: when the input value is invalid, for example `brightness=(-0.5, 0.5)`, should we raise an error or auto clip it?

Comment (fmassa) [reply torchvision/transforms/transforms.py]: the current code doesn't raise a warning, right?
Or is it handled in the underlying `F.adjust_brightness`? I think the error should be there, right?

File: torchvision/transforms/transforms.py (hunk 5 above)
Comment (fmassa) [reply torchvision/transforms/transforms.py]: What happens if brightness is set to zero in the current code? It doesn't change anything in the input, right?
In that case, why don't we just skip the transform, which will make the code a bit more efficient?

File: torchvision/transforms/transforms.py (hunk 4 above)
Comment (yaox12) [reply torchvision/transforms/transforms.py]: Only `F.adjust_hue` checks the input, brightness/contrast/saturation don't.  
The current code only need to ensure `brightness > 0`, but what if the input is `(0.5, -0.5)`, which means `min > max`, so I think we should check here.

Comment (yaox12) [reply torchvision/transforms/transforms.py]: Or should we check the input in `__init__`, just as `Line 911`(https://github.com/yaox12/vision/blob/master/torchvision/transforms/transforms.py#L911)

Comment (fmassa) [reply torchvision/transforms/transforms.py]: Let's check it in `__init__`, might be better

**Pull request merged**